Sources/FoundationEssentials/String/String+Path.swift

```diff
@@ -137,6 +137,24 @@ extension String {
     
     #if !NO_FILESYSTEM
     internal static func homeDirectoryPath(forUser user: String? = nil) -> String {
+#if os(Windows)
+        var hToken: HANDLE? = nil
+        guard OpenProcessToken(GetCurrentProcess(), TOKEN_QUERY, &hToken) else {
+            return ""
+        }
+        defer { CloseHandle(hToken) }
+
+        var dwcchSize: DWORD = 0
+        _ = GetUserProfileDirectoryW(hToken, nil, &dwcchSize)
+
+        return withUnsafeTemporaryAllocation(of: WCHAR.self, capacity: Int(dwcchSize)) {
+            var dwcchSize: DWORD = DWORD($0.count)
+            guard GetUserProfileDirectoryW(hToken, $0.baseAddress, &dwcchSize) else {
+                return ""
```

Comment:
Gotcha, I see in SCF we fallback to the process' current working directory (https://github.com/apple/swift-corelibs-foundation/blob/77af9c5a984f40e186c0ba127cff88f2a2588c15/CoreFoundation/Base.subproj/CFPlatform.c#L542) - should we do the same here? We could just return `FileManager.default.currentDirectoryPath` here.

As for the environment variables, gotcha is `USERPROFILE` a commonly used environment variable for this purpose on Windows? It looks like the SCF implementation in C use the value of `HOMEPATH` appended to `HOMEDRIVE` as a fallback if set, is it worth doing the same here?

Reply:
Hmm, that is a good fallback option. We could also consider a temporary directory.

`USERPROFILE` is the equivalent to `HOME` really. `HOMEDRIVE` is _just_ the drive component and `HOMEPATH` is the drive relative path to the user profile. I would prefer to use `USERPROFILE` as that is more "modern" (`HOMEDRIVE`, `HOMEPATH` are NT4-era variables which are no longer considered "standard") and should account for local and roaming profiles if the user is an Active Directory environment. If you reference https://learn.microsoft.com/en-us/windows/win32/shell/knownfolderid?redirectedfrom=MSDN - the known folders also reference `%UserProfile%`.

Reply:
Thanks for the explanation, in that case yeah I think using `USERPROFILE` as a potential fallback if set makes sense to me

Reply:
I believe that I have done this the best I can do.

Reply:
Thanks! Overall it looks good, but I added a comment above about using the fallback in more places. Do we want to use the current directory as the final fallback instead of the empty string too?

Reply:
That's fair, I mainly suggested it because that's what we do today in swift-corelibs-foundation - even if the program exists in a privileged location it's still entirely possible that the current working directory is a reasonable path, but I agree if we don't think that's satisfactory we don't need to replicate it here. I am concerned however about using the empty string as I feel this is prone to fail later down the line. If we have no reasonable fallback path to use, should we instead just `fatalError` here to make it clear what the problem is? Otherwise, my suspicion is that this scenario will just turn into some other function throwing because it doesn't know how to operate on an empty path

Reply:
I'm totally fine with a `fatalError`. We can always try to do alternative fallbacks if we hit these in practice.